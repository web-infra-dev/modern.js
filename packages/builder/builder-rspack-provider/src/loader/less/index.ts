import path from 'path';
import { Buffer } from 'buffer';
import { getCompiledPath } from '../../shared';

const generateOptions = (options: Less.Options): Less.Options => {
  const defaultConfig = {
    enableSourcemap: false,
  };
  return {
    ...defaultConfig,
    ...options,
  };
};

export interface Options {
  implementation?: string;
  lessOptions?: Less.Options;
}

export default async function lessLoader(loaderContext: any) {
  const meta = '';
  const { lessOptions: options, implementation } =
    loaderContext.getOptions() ?? {};

  try {
    const code = loaderContext.source.getCode();
    const final_options = generateOptions({
      filename: loaderContext.resourcePath,
      ...options,
      paths: [
        ...(options?.paths || ['node_modules']),
        path.dirname(loaderContext.resourcePath),
      ],
      plugins: [],
    });

    let lessImplementation;

    if (typeof implementation === 'string') {
      lessImplementation = require(implementation);
    } else {
      lessImplementation = require(getCompiledPath('less'));
    }
    const result = await lessImplementation.render(code, final_options);
    const { css } = result;

    return {
      content: css,
      meta: meta ? Buffer.from(JSON.stringify(meta)) : '',
    };
  } catch (error) {
    console.error(loaderContext.resourcePath);
    console.error(error);
    throw error;
  }
}
