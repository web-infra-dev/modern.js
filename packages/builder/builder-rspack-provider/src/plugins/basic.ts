import type { BuilderPlugin } from '../types';

/**
 * Provide some basic configs of rspack
 */
export const builderPluginBasic = (): BuilderPlugin => ({
  name: 'builder-plugin-basic',

  setup(api) {
    api.modifyRspackConfig(async (rspackConfig, { isProd }) => {
      // The base directory for resolving entry points and loaders from the configuration.
      rspackConfig.context = api.context.rootPath;

      rspackConfig.mode = isProd ? 'production' : 'development';

      rspackConfig.infrastructureLogging = {
        // Using `error` level to avoid `cache.PackFileCacheStrategy` logs
        level: 'error',
      };

      // todo: rspack not support configure
      // // if the chunk size exceeds 1MiB, we will throw a warning
      // chain.performance.maxAssetSize(1024 * 1024);
      // chain.performance.maxEntrypointSize(1024 * 1024);

      // // This will be futureDefaults in webpack 6
      // chain.module.parser.merge({
      //   javascript: {
      //     exportsPresence: 'error',
      //   },
      // });
    });
  },
});
