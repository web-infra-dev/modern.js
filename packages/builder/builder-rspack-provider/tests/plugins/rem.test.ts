import { describe, expect, it } from 'vitest';
import { builderPluginRem } from '@builder/plugins/rem';
import { createBuilder, matchPlugin } from '../helper';
import { builderPluginCss } from '@/plugins/css';
import { builderPluginLess } from '@/plugins/less';
import { builderPluginSass } from '@/plugins/sass';

describe('plugins/rem', () => {
  it('should not run rem plugin without config', async () => {
    const builder = await createBuilder({
      plugins: [builderPluginCss(), builderPluginRem()],
      builderConfig: {},
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();
    expect(bundlerConfigs[0]).toMatchSnapshot();
  });

  it('should not run rem plugin when false', async () => {
    const builder = await createBuilder({
      plugins: [builderPluginCss(), builderPluginRem()],
      builderConfig: {
        output: {
          convertToRem: false,
        },
      },
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();
    expect(bundlerConfigs[0]).toMatchSnapshot();
  });

  it('should run rem plugin with default config', async () => {
    const builder = await createBuilder({
      plugins: [
        builderPluginCss(),
        builderPluginLess(),
        builderPluginSass(),
        builderPluginRem(),
      ],
      builderConfig: {
        output: {
          convertToRem: true,
        },
      },
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();

    expect(bundlerConfigs[0]).toMatchSnapshot();
  });

  it('should order plugins and run rem plugin with default config', async () => {
    const builder = await createBuilder({
      plugins: [
        builderPluginRem(),
        builderPluginCss(),
        builderPluginLess(),
        builderPluginSass(),
      ],
      builderConfig: {
        output: {
          convertToRem: true,
        },
      },
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();

    expect(bundlerConfigs[0]).toMatchSnapshot();
  });

  it('should not run htmlPlugin with enableRuntime is false', async () => {
    const builder = await createBuilder({
      plugins: [builderPluginCss(), builderPluginRem()],
      builderConfig: {
        output: {
          convertToRem: {
            enableRuntime: false,
          },
        },
      },
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();

    expect(bundlerConfigs[0].plugins?.length || 0).toBe(0);
    expect(bundlerConfigs[0]).toMatchSnapshot();
  });

  it('should run rem plugin with custom config', async () => {
    const builder = await createBuilder({
      plugins: [builderPluginCss(), builderPluginRem()],
      builderConfig: {
        output: {
          convertToRem: {
            rootFontSize: 30,
            pxtorem: {
              propList: ['font-size'],
            },
          },
        },
      },
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();
    expect(bundlerConfigs[0]).toMatchSnapshot();
  });

  it('should not run rem plugin when target is node', async () => {
    const builder = await createBuilder({
      plugins: [builderPluginCss(), builderPluginRem()],
      builderConfig: {
        output: {
          convertToRem: true,
        },
      },
      target: ['node'],
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();

    expect(
      matchPlugin(bundlerConfigs[0], 'AutoSetRootFontSizePlugin'),
    ).toBeFalsy();
  });

  it('should not run rem plugin when target is web-worker', async () => {
    const builder = await createBuilder({
      plugins: [builderPluginCss(), builderPluginRem()],
      builderConfig: {
        output: {
          convertToRem: true,
        },
      },
      target: ['web-worker'],
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();

    expect(
      matchPlugin(bundlerConfigs[0], 'AutoSetRootFontSizePlugin'),
    ).toBeFalsy();
  });
});
