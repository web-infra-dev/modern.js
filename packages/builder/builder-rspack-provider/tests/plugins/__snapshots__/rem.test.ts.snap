// Vitest Snapshot v1

exports[`plugins/rem > should not run htmlPlugin with enableRuntime is false 1`] = `
{
  "builtins": {
    "css": {
      "modules": {
        "exportsOnly": false,
        "localIdentName": "[path][name]__[local]--[hash:5]",
        "localsConvention": "camelCase",
      },
    },
  },
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "sideEffects": true,
            "test": /\\\\\\.module\\\\\\.\\\\w\\+\\$/i,
            "type": "css/module",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "*",
                    ],
                    "rootValue": 50,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
          {
            "sideEffects": true,
            "type": "css",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "*",
                    ],
                    "rootValue": 50,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
        ],
        "test": /\\\\\\.css\\$/,
      },
    ],
  },
}
`;

exports[`plugins/rem > should not run rem plugin when false 1`] = `
{
  "builtins": {
    "css": {
      "modules": {
        "exportsOnly": false,
        "localIdentName": "[path][name]__[local]--[hash:5]",
        "localsConvention": "camelCase",
      },
    },
  },
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "sideEffects": true,
            "test": /\\\\\\.module\\\\\\.\\\\w\\+\\$/i,
            "type": "css/module",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
          {
            "sideEffects": true,
            "type": "css",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
        ],
        "test": /\\\\\\.css\\$/,
      },
    ],
  },
}
`;

exports[`plugins/rem > should not run rem plugin without config 1`] = `
{
  "builtins": {
    "css": {
      "modules": {
        "exportsOnly": false,
        "localIdentName": "[path][name]__[local]--[hash:5]",
        "localsConvention": "camelCase",
      },
    },
  },
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "sideEffects": true,
            "test": /\\\\\\.module\\\\\\.\\\\w\\+\\$/i,
            "type": "css/module",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
          {
            "sideEffects": true,
            "type": "css",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
        ],
        "test": /\\\\\\.css\\$/,
      },
    ],
  },
}
`;

exports[`plugins/rem > should order plugins and run rem plugin with default config 1`] = `
{
  "builtins": {
    "css": {
      "modules": {
        "exportsOnly": false,
        "localIdentName": "[path][name]__[local]--[hash:5]",
        "localsConvention": "camelCase",
      },
    },
  },
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "sideEffects": true,
            "test": /\\\\\\.module\\\\\\.\\\\w\\+\\$/i,
            "type": "css/module",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "*",
                    ],
                    "rootValue": 50,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
          {
            "sideEffects": true,
            "type": "css",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "*",
                    ],
                    "rootValue": 50,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
        ],
        "test": /\\\\\\.css\\$/,
      },
      {
        "oneOf": [
          {
            "sideEffects": true,
            "test": /\\\\\\.module\\\\\\.\\\\w\\+\\$/i,
            "type": "css/module",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "*",
                    ],
                    "rootValue": 50,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
              {
                "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/less-loader",
                "options": {
                  "implementation": "<WORKSPACE>/packages/builder/builder-shared/compiled/less",
                  "lessOptions": {
                    "javascriptEnabled": true,
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
          {
            "sideEffects": true,
            "type": "css",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "*",
                    ],
                    "rootValue": 50,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
              {
                "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/less-loader",
                "options": {
                  "implementation": "<WORKSPACE>/packages/builder/builder-shared/compiled/less",
                  "lessOptions": {
                    "javascriptEnabled": true,
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
        ],
        "test": /\\\\\\.less\\$/,
      },
      {
        "oneOf": [
          {
            "sideEffects": true,
            "test": /\\\\\\.module\\\\\\.\\\\w\\+\\$/i,
            "type": "css/module",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "*",
                    ],
                    "rootValue": 50,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
              {
                "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/sass-loader",
                "options": {
                  "implementation": "<WORKSPACE>/packages/builder/builder-shared/compiled/sass",
                  "sourceMap": true,
                },
              },
            ],
          },
          {
            "sideEffects": true,
            "type": "css",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "*",
                    ],
                    "rootValue": 50,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
              {
                "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/sass-loader",
                "options": {
                  "implementation": "<WORKSPACE>/packages/builder/builder-shared/compiled/sass",
                  "sourceMap": true,
                },
              },
            ],
          },
        ],
        "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
      },
    ],
  },
  "plugins": [
    AutoSetRootFontSizePlugin {
      "HtmlPlugin": [Function],
      "distDir": "static/js",
      "name": "AutoSetRootFontSizePlugin",
      "options": {
        "enableRuntime": true,
        "excludeEntries": [],
        "inlineRuntime": true,
        "maxRootFontSize": 64,
        "rootFontSize": 50,
        "rootFontSizeVariableName": "ROOT_FONT_SIZE",
        "screenWidth": 375,
        "supportLandscape": false,
        "useRootFontSizeBeyondMax": false,
        "widthQueryKey": "",
      },
      "scriptPath": "",
      "webpackEntries": [],
    },
  ],
}
`;

exports[`plugins/rem > should run rem plugin with custom config 1`] = `
{
  "builtins": {
    "css": {
      "modules": {
        "exportsOnly": false,
        "localIdentName": "[path][name]__[local]--[hash:5]",
        "localsConvention": "camelCase",
      },
    },
  },
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "sideEffects": true,
            "test": /\\\\\\.module\\\\\\.\\\\w\\+\\$/i,
            "type": "css/module",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "font-size",
                    ],
                    "rootValue": 30,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
          {
            "sideEffects": true,
            "type": "css",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "font-size",
                    ],
                    "rootValue": 30,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
        ],
        "test": /\\\\\\.css\\$/,
      },
    ],
  },
  "plugins": [
    AutoSetRootFontSizePlugin {
      "HtmlPlugin": [Function],
      "distDir": "static/js",
      "name": "AutoSetRootFontSizePlugin",
      "options": {
        "enableRuntime": true,
        "excludeEntries": [],
        "inlineRuntime": true,
        "maxRootFontSize": 64,
        "pxtorem": {
          "propList": [
            "font-size",
          ],
        },
        "rootFontSize": 30,
        "rootFontSizeVariableName": "ROOT_FONT_SIZE",
        "screenWidth": 375,
        "supportLandscape": false,
        "useRootFontSizeBeyondMax": false,
        "widthQueryKey": "",
      },
      "scriptPath": "",
      "webpackEntries": [],
    },
  ],
}
`;

exports[`plugins/rem > should run rem plugin with default config 1`] = `
{
  "builtins": {
    "css": {
      "modules": {
        "exportsOnly": false,
        "localIdentName": "[path][name]__[local]--[hash:5]",
        "localsConvention": "camelCase",
      },
    },
  },
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "sideEffects": true,
            "test": /\\\\\\.module\\\\\\.\\\\w\\+\\$/i,
            "type": "css/module",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "*",
                    ],
                    "rootValue": 50,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
          {
            "sideEffects": true,
            "type": "css",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "*",
                    ],
                    "rootValue": 50,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
        ],
        "test": /\\\\\\.css\\$/,
      },
      {
        "oneOf": [
          {
            "sideEffects": true,
            "test": /\\\\\\.module\\\\\\.\\\\w\\+\\$/i,
            "type": "css/module",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "*",
                    ],
                    "rootValue": 50,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
              {
                "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/less-loader",
                "options": {
                  "implementation": "<WORKSPACE>/packages/builder/builder-shared/compiled/less",
                  "lessOptions": {
                    "javascriptEnabled": true,
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
          {
            "sideEffects": true,
            "type": "css",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "*",
                    ],
                    "rootValue": 50,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
              {
                "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/less-loader",
                "options": {
                  "implementation": "<WORKSPACE>/packages/builder/builder-shared/compiled/less",
                  "lessOptions": {
                    "javascriptEnabled": true,
                  },
                  "sourceMap": true,
                },
              },
            ],
          },
        ],
        "test": /\\\\\\.less\\$/,
      },
      {
        "oneOf": [
          {
            "sideEffects": true,
            "test": /\\\\\\.module\\\\\\.\\\\w\\+\\$/i,
            "type": "css/module",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "*",
                    ],
                    "rootValue": 50,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
              {
                "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/sass-loader",
                "options": {
                  "implementation": "<WORKSPACE>/packages/builder/builder-shared/compiled/sass",
                  "sourceMap": true,
                },
              },
            ],
          },
          {
            "sideEffects": true,
            "type": "css",
            "use": [
              {
                "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rspack/postcss-loader/src/index.js",
                "options": {
                  "postcssOptions": {
                    "plugins": [
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      [Function],
                      {
                        "browsers": [
                          "> 0.01%",
                          "not dead",
                          "not op_mini all",
                        ],
                        "info": [Function],
                        "options": {
                          "flexbox": "no-2009",
                          "overrideBrowserslist": [
                            "> 0.01%",
                            "not dead",
                            "not op_mini all",
                          ],
                        },
                        "postcssPlugin": "autoprefixer",
                        "prepare": [Function],
                      },
                    ],
                  },
                  "pxToRem": {
                    "propList": [
                      "*",
                    ],
                    "rootValue": 50,
                    "unitPrecision": 5,
                  },
                  "sourceMap": true,
                },
              },
              {
                "loader": "<WORKSPACE>/packages/builder/builder-shared/compiled/sass-loader",
                "options": {
                  "implementation": "<WORKSPACE>/packages/builder/builder-shared/compiled/sass",
                  "sourceMap": true,
                },
              },
            ],
          },
        ],
        "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
      },
    ],
  },
  "plugins": [
    AutoSetRootFontSizePlugin {
      "HtmlPlugin": [Function],
      "distDir": "static/js",
      "name": "AutoSetRootFontSizePlugin",
      "options": {
        "enableRuntime": true,
        "excludeEntries": [],
        "inlineRuntime": true,
        "maxRootFontSize": 64,
        "rootFontSize": 50,
        "rootFontSizeVariableName": "ROOT_FONT_SIZE",
        "screenWidth": 375,
        "supportLandscape": false,
        "useRootFontSizeBeyondMax": false,
        "widthQueryKey": "",
      },
      "scriptPath": "",
      "webpackEntries": [],
    },
  ],
}
`;
